Sources/CancelContext.swift

```diff
@@ -0,0 +1,270 @@
+    }
+    
+    private func crossCancel(childContext: CancelContext, parentCancelledError: Error?) {
+        let parentError = parentCancelledError
+        let childError =  childContext.cancelledError
+        
+        if parentError != nil {
+            if childError == nil {
+                childContext.cancel(error: parentError)
+            }
+        } else if childError != nil {
+            if parentError == nil {
+                cancel(error: childError)
+            }
+        }
+    }
+    
+    func recover() {
+        cancelledError = nil
+    }
+    
+    func removeItems(_ list: CancelItemList, clearList: Bool) -> Error? {
+        var error: Error?
+        barrier.sync(flags: .barrier) {
+            error = internalCancelledError
+            if error == nil && list.items.count != 0 {
+                var currentIndex = 1
+                // The `list` parameter should match a block of items in the cancelItemList, remove them from the cancelItemList
+                // in one operation for efficiency
+                if cancelItemSet.remove(list.items[0]) != nil {
+                    let removeIndex = cancelItems.index(of: list.items[0])!
+                    while currentIndex < list.items.count {
+                        let item = list.items[currentIndex]
+                        if item != cancelItems[removeIndex + currentIndex] {
+                            break
+                        }
+                        cancelItemSet.remove(item)
+                        currentIndex += 1
+                    }
+                    cancelItems.removeSubrange(removeIndex..<(removeIndex+currentIndex))
+                }
+                
+                // Remove whatever falls outside of the block
+                while currentIndex < list.items.count {
+                    let item = list.items[currentIndex]
+                    if cancelItemSet.remove(item) != nil {
+                        cancelItems.remove(at: cancelItems.index(of: item)!)
+                    }
+                    currentIndex += 1
+                }
+                
+                if clearList {
+                    list.removeAll()
+                }
+            }
+        }
+        return error
+    }
+}
+
+/// Tracks the cancel items for a CancellablePromise.  These items are removed from the associated CancelContext when the promise resolves.
+public class CancelItemList {
+    fileprivate var items: [CancelItem]
+    
+    init() {
+        self.items = []
+    }
+    
+    func append(_ item: CancelItem) {
+        items.append(item)
+    }
+    
+    func append(contentsOf list: CancelItemList, clearList: Bool) {
+        items.append(contentsOf: list.items)
+        if clearList {
+            list.removeAll()
+        }
+    }
+    
+    func removeAll() {
+        items.removeAll()
+    }
+}
+
+class CancelItem: Hashable {
+    lazy var hashValue: Int = {
```

Comment:
`lazy` is not thread-safe.

Reply:
Wow, great catch!  The `CancelContext` instance's `hashValue` could indeed be accessed by multiple threads at the same time.

I fixed `CancelContext` by making a `Hashable` wrapper class, where the wrapper instance will only be accessed by a single thread.  Other threads make their own wrapper instances so this solves the thread safety issue with `CancelContext`.

`CancelItem` also has a `lazy` `hashValue`, but all dictionary operations involving `CancelItem` are protected by the `barrier` lock.

Admittedly this is a slightly convoluted.  But I want to use `ObjectIdentifier.hashValue` for the `CancelContext` and `CancelItem` hash values, which then forces you to use `lazy`.

I'm very open to ideas for simplifying this, lmk if you think of anything!!